File: python/luhn/luhn.py
class Luhn:
    def __init__(self, candidate):
        self.candidate = candidate.replace(' ', '')

    def valid(self):
        if len(self.candidate) <= 1 or not self.candidate.isdigit():
            return False

        return sum(
            self.convert(len(self.candidate) - 1 - i, int(self.candidate[i]))
            for i in range(len(self.candidate))
        ) % 10 == 0

    def convert(self, right_index, value):
        return self.double(value) if right_index % 2 != 0 else value

    def double(self, x):
        result = x * 2
        if result > 9:
            result -= 9
        return result

File: cpp/luhn/luhn.cpp
#include "luhn.h"
#include <iostream>

namespace luhn
{
bool valid(std::string s)
{
    s.erase(std::remove(s.begin(), s.end(), ' '), s.end());

    if (s.length() <= 1 || !std::all_of(s.begin(), s.end(), [](char ch) { return std::isdigit(ch); }))
    {
        return false;
    }

    int sum{0};
    for (std::string::size_type i{0}; i < s.length(); ++i)
    {
        int value{s[i] - '0'};
        if ((s.length() - 1 - i) % 2)
        {
            value *= 2;
            if (value > 9)
            {
                value -= 9;
            }
        }

        sum += value;
    }

    return sum % 10 == 0;
}
} // namespace luhn
